java-checks/src/main/java/org/sonar/java/checks/TypeParametersShadowingCheck.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2021 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+import org.sonar.check.Rule;
+import org.sonar.plugins.java.api.JavaFileScanner;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.TypeParameters;
+
+@Rule(key = "S4977")
+public class TypeParametersShadowingCheck extends BaseTreeVisitor implements JavaFileScanner {
+
+  private static final String ISSUE_MESSAGE = "Rename \"%s\" which hides a type parameter from the outer scope.";
+
+  private JavaFileScannerContext context;
+
+  private Map<String, IdentifierTree> currentTypeParametersInScope;
+  private Deque<Map<String, IdentifierTree>> scopes;
+
+  @Override
+  public void scanFile(JavaFileScannerContext context) {
+    this.context = context;
+    scopes = new ArrayDeque<>();
+    currentTypeParametersInScope = new HashMap<>();
+    scan(context.getTree());
+  }
+
+  @Override
+  public void visitClass(ClassTree tree) {
+    processTree(tree, tree.typeParameters(), tree.symbol().isStatic(), super::visitClass);
+  }
+
+  @Override
+  public void visitMethod(MethodTree tree) {
+    processTree(tree, tree.typeParameters(), tree.symbol().isStatic(), super::visitMethod);
+  }
+
+  private <T> void processTree(T tree, TypeParameters typeParameters, boolean isStatic, Consumer<T> visitTree) {
+    if (isStatic) {
+      pushNewScope();
+    }
+    List<String> declaredTypeParameters = processAndGetTypeParameters(typeParameters);
+    visitTree.accept(tree);
+    declaredTypeParameters.forEach(currentTypeParametersInScope::remove);

```

Comment:
I think it would be better if the new entries were put into `currentTypeParametersInScope` here instead of in `processAndTypeParameters`:

    List<String> declaredTypeParameters = processAndGetTypeParameters(typeParameters);
    currentTypeParamtersInScope.putAll(declaredTypeParameters);
    visitTree.accept(tree);
    declaredTypeParameters.forEach(currentTypeParametersInScope::remove);

This way the logic of "add entries, visit children, remove entries" becomes more apparent and the addition step and the remove step are more symmetric.